File: src/ros2_bridge/mappers/sensor_msgs/imu.rs
//! Typed mapper for `sensor_msgs/msg/Imu`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn imu_to_bus(msg: ros_env::sensor_msgs::msg::Imu) -> crate::sensor_msgs::msg::v1::Imu {
    crate::sensor_msgs::msg::v1::Imu {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        orientation: Some(
            crate::ros2_bridge::mappers::geometry_msgs::quaternion::quaternion_to_bus(
                msg.orientation,
            ),
        ),
        orientation_covariance: crate::ros2_bridge::mappers::convert::f64_seq(
            msg.orientation_covariance,
        ),
        angular_velocity: Some(
            crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(
                msg.angular_velocity,
            ),
        ),
        angular_velocity_covariance: crate::ros2_bridge::mappers::convert::f64_seq(
            msg.angular_velocity_covariance,
        ),
        linear_acceleration: Some(
            crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(
                msg.linear_acceleration,
            ),
        ),
        linear_acceleration_covariance: crate::ros2_bridge::mappers::convert::f64_seq(
            msg.linear_acceleration_covariance,
        ),
    }
}

pub(crate) fn imu_to_ros(bus: crate::sensor_msgs::msg::v1::Imu) -> ros_env::sensor_msgs::msg::Imu {
    ros_env::sensor_msgs::msg::Imu {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(
            bus.header.unwrap_or_default(),
        ),
        orientation: crate::ros2_bridge::mappers::geometry_msgs::quaternion::quaternion_to_ros(
            bus.orientation.unwrap_or_default(),
        ),
        orientation_covariance: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(
            bus.orientation_covariance,
        ),
        angular_velocity: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(
            bus.angular_velocity.unwrap_or_default(),
        ),
        angular_velocity_covariance: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(
            bus.angular_velocity_covariance,
        ),
        linear_acceleration: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(
            bus.linear_acceleration.unwrap_or_default(),
        ),
        linear_acceleration_covariance:
            crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(
                bus.linear_acceleration_covariance,
            ),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsImuMapper;

impl TypedTopicMapper for SensorMsgsImuMapper {
    type Ros = ros_env::sensor_msgs::msg::Imu;
    type Bus = crate::sensor_msgs::msg::v1::Imu;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(imu_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(imu_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/imu.py
"""Generated mapper for `sensor_msgs/msg/Imu`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.quaternion import quaternion_to_bus, quaternion_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.vector3 import vector3_to_bus, vector3_to_ros

def imu_to_bus(msg):
    from robot_bus.sensor_msgs.msg.v1 import Imu as BusMsg

    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.orientation.CopyFrom(quaternion_to_bus(msg.orientation))
    bus.orientation_covariance.extend(list(msg.orientation_covariance))
    bus.angular_velocity.CopyFrom(vector3_to_bus(msg.angular_velocity))
    bus.angular_velocity_covariance.extend(list(msg.angular_velocity_covariance))
    bus.linear_acceleration.CopyFrom(vector3_to_bus(msg.linear_acceleration))
    bus.linear_acceleration_covariance.extend(list(msg.linear_acceleration_covariance))
    return bus


def imu_to_ros(bus):
    from sensor_msgs.msg import Imu as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.orientation = quaternion_to_ros(bus.orientation)
    out.orientation_covariance = list(bus.orientation_covariance)
    out.angular_velocity = vector3_to_ros(bus.angular_velocity)
    out.angular_velocity_covariance = list(bus.angular_velocity_covariance)
    out.linear_acceleration = vector3_to_ros(bus.linear_acceleration)
    out.linear_acceleration_covariance = list(bus.linear_acceleration_covariance)
    return out


class SensorMsgsImuMapper:
    def type_name(self) -> str:
        return "sensor_msgs/msg/Imu"

    def ros_msg_type(self):
        from sensor_msgs.msg import Imu as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return imu_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.sensor_msgs.msg.v1 import Imu as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return imu_to_ros(bus)
